### PR TITLE
[Labs] Phox mmd loss function

### PR DESCRIPTION
**Pull request by austingmhuang**
**Context:**
We implement mmd_loss for labs

**Description of the Change:**
Implemented mmd_loss and a few helper functions, plus renamed some variables

**Benefits:**
Now has mmd_loss

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: pennylane/labs/tests/phox/test_mmd_loss.py [hunk 1]
```diff
@@ -0,0 +1,591 @@
...
+            # 2-qubit, 3-gate circuit
+            (
+                [[0], [1], [0, 1]],
+                [0.2, 0.8, 0.4],
+                [0.4, 0.6],
+                100,
+            ),
+        ],
+    )
+    def test_unbiased_z_test(self, generators, params, biases, n_data):
```
Comment (austingmhuang) [inline pennylane/labs/tests/phox/test_mmd_loss.py:488]: @josephbowles I think this test should be verify the correctness of our mmd_loss function.

File: pennylane/labs/tests/phox/test_mmd_loss.py [hunk 2]
```diff
@@ -0,0 +1,591 @@
...
+    1. Compute the *exact* MMD^2 from PennyLane probabilities and data.
+    2. Collect many stochastic phox estimates using different PRNG keys
+       and random data sub-samples.
+    3. Verify the sample mean is consistent with the exact value via a
+       two-sided Z-test:  ``|exact - mean_est| / SE < z_threshold``.
+    """
+
+    Z_THRESHOLD = 4.0
+    N_TRIALS = 30
+    N_OPS = 400
```
Comment (josephbowles) [inline pennylane/labs/tests/phox/test_mmd_loss.py]: might be worth cranking up N_OPS and N_SAMPLES to be more sure the central limit theorem is a good approximation. Perhaps N_OPS=5000, N_SAMPLES=5000? Or do you think that would be too much for a test?

File: pennylane/labs/tests/phox/test_mmd_loss.py [hunk 3]
```diff
@@ -0,0 +1,591 @@
...
+
+        sigma = float(median_heuristic(X))
+        exact = _exact_mmd2_kernel(probs_p, X, sigma, n_qubits)
+
+        gates = {i: [w] for i, w in enumerate(generators)}
+        batch = min(80, n_data)
+
+        estimates = []
+        for t in range(self.N_TRIALS):
+            key = jax.random.PRNGKey(t * 17 + 7)
```
Comment (josephbowles) [inline pennylane/labs/tests/phox/test_mmd_loss.py]: why t*17 +7?

Comment (josephbowles) [reply pennylane/labs/tests/phox/test_mmd_loss.py]: because it is a sort of random thing that avoids relying on jax.random.split?

Comment (austingmhuang) [reply pennylane/labs/tests/phox/test_mmd_loss.py]: It's a holdover from my notebook when I wasn't using jax.random.split and I wanted a key that wouldn't accidentally collide with something. This should use `jax.random.split` instead.

Comment (austingmhuang) [inline pennylane/labs/tests/phox/test_mmd_loss.py]:
```suggestion
        root_key = jax.random.PRNGKey(42)
        trial_keys = jax.random.split(root_key, self.N_TRIALS)
        
        estimates = []
        for t in range(self.N_TRIALS):
            key = trial_keys[t]
```
should've been what I wrote

File: pennylane/labs/tests/phox/test_mmd_loss.py (hunk 2 above)
Comment (austingmhuang) [reply pennylane/labs/tests/phox/test_mmd_loss.py]: The main thing is that it ends up taking a pretty long time and I'm actually not sure how labs tests run in the CI and if it will make the CI unreasonably slow. I can crank it up locally and get passing tests (after some wait). Probably better to play it safe and not accidentally blow up the CI.

**Review by willjmax — APPROVED** [review]

**Review by josephbowles — APPROVED** [review]
